cpp/include/cudf/binning/bin.cuh

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+// TODO: Clean up includes before pushing a final version.
+#include <cudf/column/column_view.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/binning/bin.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/types.hpp>
+#include <thrust/functional.h>
+#include <thrust/binary_search.h>
+#include <thrust/execution_policy.h>
+#include <cudf/copying.hpp>
+
+namespace cudf {
+
+namespace bin {
+
+// TODO: This is a placeholder to remind myself to figure out the proper way to
+// set nulls. I need to modify the bitmask, but even if I passed a pointer to
+// it directly to the constructor of bin_finder and modified it directly (which
+// would be a terribly hacky solution), the current index is not provided to
+// the call operator so I wouldn't know what bit to set (and in any case that
+// would not be parallel-safe without atomics).
+constexpr unsigned int MYNULL = 0xffffffff;
```

Comment:
I think using a sentinel value for the bin label is a good solution. Then you can use `valid_if` to materialize a bitmask corresponding to the sentinel values.

https://github.com/rapidsai/cudf/blob/c929ba1fe85c152d6e8b4c868cd36f0802dafa51/cpp/include/cudf/detail/valid_if.cuh#L94